### PR TITLE
Cancellation of lobby selection

### PR DESCRIPTION
**Pull request by RAIXOCZ**
+ Allow cancellation of lobby choose event
+ Adding a function to check if an event is triggered because of a fallback server selection

### CODE REVIEW

File: API/src/main/java/xyz/kyngs/librelogin/api/event/events/LobbyServerChooseEvent.java [hunk 1]
```diff
@@ -15,4 +15,26 @@
...
+     * @param cancelled If set to true, this event will be aborted
+     */
+    void setCancelled(boolean cancelled);
+
+    /**
+     * Checks if the event is aborted
+     *
+     * @return Whether is the event aborted
+     * */
+    boolean isCancelled();
```
Comment (kyngs) [inline API/src/main/java/xyz/kyngs/librelogin/api/event/events/LobbyServerChooseEvent.java:38]: Perhaps the canceling of an event should be defined by a super class such as CancellableEvent?

File: API/src/main/java/xyz/kyngs/librelogin/api/server/ServerHandler.java [hunk 2]
```diff
@@ -37,6 +49,7 @@ public interface ServerHandler<P, S> {
      * @param remember Whether to respect the remember last server option
      * @return An optimal lobby server, or null if there are no lobby servers
      */
+    @Deprecated
```
Comment (kyngs) [inline API/src/main/java/xyz/kyngs/librelogin/api/server/ServerHandler.java:52]: What's the reason to use optional here? Deprecating an API just to move it to Optional doesn't make sense.

**Review by kyngs — CHANGES_REQUESTED** [review]
This PR goes out of scope way more than it should've. Migrating to Optional for one method doesn't make sense.

File: API/src/main/java/xyz/kyngs/librelogin/api/event/events/LobbyServerChooseEvent.java (hunk 1 above)
Comment (RAIXOCZ) [reply API/src/main/java/xyz/kyngs/librelogin/api/event/events/LobbyServerChooseEvent.java]: Fixed this in [6824b00](https://github.com/kyngs/LibreLogin/pull/103/commits/6824b0025b2acbb09b6fd3bf686d77c1018cd577)

File: API/src/main/java/xyz/kyngs/librelogin/api/server/ServerHandler.java (hunk 2 above)
Comment (RAIXOCZ) [reply API/src/main/java/xyz/kyngs/librelogin/api/server/ServerHandler.java]: Fixed this in [6824b00](https://github.com/kyngs/LibreLogin/pull/103/commits/6824b0025b2acbb09b6fd3bf686d77c1018cd577)

File: API/src/main/java/xyz/kyngs/librelogin/api/server/ServerHandler.java [hunk 3]
```diff
@@ -11,6 +11,7 @@
 import xyz.kyngs.librelogin.api.database.User;
 
 import java.util.Collection;
+import java.util.Optional;
```
Comment (kyngs) [inline API/src/main/java/xyz/kyngs/librelogin/api/server/ServerHandler.java:14]: Unused import?

File: API/src/main/java/xyz/kyngs/librelogin/api/server/ServerHandler.java [hunk 4]
```diff
@@ -29,6 +30,17 @@ public interface ServerHandler<P, S> {
...
+    /**
+     * Chooses an optimal lobby server to connect the player to. Usually the one with the lowest player count.
+     *
+     * @param user     The user of the player, or null if the player is from bedrock
+     * @param player   The player we're choosing the server for
+     * @param remember Whether to respect the remember last server option
+     * @param fallback Whether to select fallback server or not
+     * @return An optimal lobby server, or null if there are no lobby servers
+     */
+    S chooseLobbyServer(@Nullable User user, P player, boolean remember, boolean fallback);
```
Comment (kyngs) [inline API/src/main/java/xyz/kyngs/librelogin/api/server/ServerHandler.java:42]: The return suggests that if there are no lobbies, null is returned. 
However, the implementation does the opposite:
If there are no lobbies, NoSuchElementException is thrown. If the event was canceled, null is returned. 

Perhaps some unchecked exception like CancelledEventException shall be thrown and documented to indicate that the event was canceled? In theory, this is a breaking change but shouldn't occur unless some other plugin cancels the event.

**Review by kyngs — CHANGES_REQUESTED** [review]

**Comment by kyngs** [comment]
Also, the license header is not present in the newly added class, you can add it using ./gradlew updateLicenses

File: Plugin/src/main/java/xyz/kyngs/librelogin/common/server/AuthenticServerHandler.java [hunk 5]
```diff
@@ -129,7 +128,12 @@ public S chooseLobbyServer(@Nullable User user, P player, boolean remember) {
...
                 })
                 .min(Comparator.comparingInt(o -> plugin.getPlatformHandle().getConnectedPlayers(o)))
-                .orElse(null);
+                .orElseThrow();
+    }
+
+    @Override
+    public S chooseLobbyServer(@Nullable User user, P player, boolean remember) {
+        return chooseLobbyServer(user, player, remember, false);
     }
```
[the remark this answers is not in the document]
Comment (RAIXOCZ) [reply Plugin/src/main/java/xyz/kyngs/librelogin/common/server/AuthenticServerHandler.java]: Fixed this in [9e7d402](https://github.com/kyngs/LibreLogin/pull/103/commits/9e7d4023429dcf904d03cd0887faf67d7ffdbfb2)

File: API/src/main/java/xyz/kyngs/librelogin/api/server/ServerHandler.java (hunk 3 above)
Comment (RAIXOCZ) [reply API/src/main/java/xyz/kyngs/librelogin/api/server/ServerHandler.java]: Fixed this in [9e7d402](https://github.com/kyngs/LibreLogin/pull/103/commits/9e7d4023429dcf904d03cd0887faf67d7ffdbfb2)

File: API/src/main/java/xyz/kyngs/librelogin/api/server/ServerHandler.java (hunk 4 above)
Comment (RAIXOCZ) [reply API/src/main/java/xyz/kyngs/librelogin/api/server/ServerHandler.java:43]: Fixed this in [9e7d402](https://github.com/kyngs/LibreLogin/pull/103/commits/9e7d4023429dcf904d03cd0887faf67d7ffdbfb2)

File: API/src/main/java/xyz/kyngs/librelogin/api/server/ServerHandler.java [hunk 6]
```diff
@@ -29,6 +29,19 @@ public interface ServerHandler<P, S> {
...
     ServerPing getLatestPing(S server);
 
+    /**
+     * Chooses an optimal lobby server to connect the player to. Usually the one with the lowest player count.
+     *
+     * @param user     The user of the player, or null if the player is from bedrock
+     * @param player   The player we're choosing the server for
+     * @param remember Whether to respect the remember last server option
+     * @param fallback Whether to select fallback server or not
+     * @throws java.util.NoSuchElementException If there is no lobby available
```
Comment (kyngs) [inline API/src/main/java/xyz/kyngs/librelogin/api/server/ServerHandler.java:39]: Why throw NoSuchElementException if the original behavior was to return null? This is an unnecessary breaking change.

File: API/src/main/java/xyz/kyngs/librelogin/api/server/ServerHandler.java [hunk 7]
```diff
@@ -29,6 +29,19 @@ public interface ServerHandler<P, S> {
...
+     * @throws xyz.kyngs.librelogin.api.event.exception.EventCancelledException If lobby choose event is cancelled
```
Comment (kyngs) [inline API/src/main/java/xyz/kyngs/librelogin/api/server/ServerHandler.java:40]: Also, these throws should be added to the method signature and also to the signature and docs of the other method.

**Review by kyngs — CHANGES_REQUESTED** [review]

File: API/src/main/java/xyz/kyngs/librelogin/api/server/ServerHandler.java (hunk 7 above)
Comment (RAIXOCZ) [reply API/src/main/java/xyz/kyngs/librelogin/api/server/ServerHandler.java:40]: Fixed this in [52d3deb](https://github.com/kyngs/LibreLogin/pull/103/commits/52d3deba0498437031f935272e925f3b95bccb65)

File: API/src/main/java/xyz/kyngs/librelogin/api/server/ServerHandler.java (hunk 6 above)
Comment (RAIXOCZ) [reply API/src/main/java/xyz/kyngs/librelogin/api/server/ServerHandler.java]: Fixed this in [52d3deb](https://github.com/kyngs/LibreLogin/pull/103/commits/52d3deba0498437031f935272e925f3b95bccb65)

**Review by kyngs — APPROVED** [review]

**Pull request merged**